SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoopInvariantCodeMotion.swift

```diff
@@ -0,0 +1,1020 @@
+
+        newExitBlocks.push(succesor)
+
+        if succesor.instructions.contains(where: { isIdenticalTo($0) }) {
+          canMove = false
+        } else if canMove {
+          move(before: succesor.instructions.first!, context)
+        } else if let firstInstruction = succesor.instructions.first {
+          copy(before: firstInstruction, context)
+        } else {
+          continue
+        }
+
+        changed = true
+      }
+    }
+
+    if changed && !canMove {
+      context.erase(instruction: self)
+    }
+
+    return changed
+  }
+  
+  /// Returns `true` if `sideEffect` cannot be reordered with a call to this
+  /// global initializer.
+  private func globalInitMayConflictWith(
+    sideEffect: Instruction,
+    _ aliasAnalysis: AliasAnalysis
+  ) -> Bool {
+    switch sideEffect {
+    case let storeInst as StoreInst:
+      return mayReadOrWrite(address: storeInst.destinationOperand.value, aliasAnalysis)
+    case let loadInst as LoadInst:
+      return mayWrite(toAddress: loadInst.operand.value, aliasAnalysis)
+    case is CondFailInst:
+      return false
+    default:
+      return true
+    }
+  }
+  
+  /// Returns `true` if any of the instructions in `sideEffects` cannot be
+  /// reordered with a call to this global initializer (which is in the same basic
+  /// block).
+  func globalInitMayConflictWith(
+    blockSideEffectSegment: StackWithCount<Instruction>.Segment,
+    _ aliasAnalysis: AliasAnalysis
+  ) -> Bool {
+    return blockSideEffectSegment
+      .contains { sideEffect in
+        globalInitMayConflictWith(
+          sideEffect: sideEffect,
+          aliasAnalysis
+        )
+      }
+  }
+
+  /// Returns `true` if any of the instructions in `loopSideEffects` which are
+  /// post-dominated by a call to this global initializer cannot be reordered with
+  /// the call.
+  func globalInitMayConflictWith(
+    loopSideEffects: StackWithCount<Instruction>,
+    notPostDominatingPreheader: BasicBlock,
+    _ aliasAnalysis: AliasAnalysis,
+    _ postDomTree: PostDominatorTree
+  ) -> Bool {
+    guard parentBlock.postDominates(notPostDominatingPreheader, postDomTree) else {
+      return true
+    }
+
+    return loopSideEffects
+      .contains { sideEffect in
+        // Only check instructions in blocks which are "before" (i.e. post-dominated
+        // by) the block which contains the init-call.
+        // Instructions which are before the call in the same block have already
+        // been checked.
+        parentBlock.strictlyPostDominates(
+          sideEffect.parentBlock,
+          postDomTree
+        ) && globalInitMayConflictWith(
+          sideEffect: sideEffect,
+          aliasAnalysis
+        )
+      }
+  }
+  
+  func isInAccessScope(of beginAccess: BeginAccessInst, _ domTree: DominatorTree) -> Bool {
+    return beginAccess.dominates(self, domTree) && beginAccess.endAccessInstructions
+      .allSatisfy { endAccessInst in
+        self.dominates(endAccessInst, domTree)
```

Comment:
This is wrong. You cannot use dominance to check if something is within a life range. That must be done with a dataflow algorithm